framework/src/relationshipmanagers/GhostBoundary.C

```diff
@@ -0,0 +1,135 @@
+  // need to delete remote elements after the addition of late geometric ghosting functors
+  // (including this ghosting functor)
+  return params;
+}
+
+GhostBoundary::GhostBoundary(const InputParameters & params)
+  : RelationshipManager(params),
+    _boundary_name(getParam<std::vector<BoundaryName>>("boundary")),
+    _is_coupling_functor(isType(Moose::RelationshipManagerType::GEOMETRIC)),
+    _ghost_point_neighbors(getParam<bool>("ghost_point_neighbors"))
+{
+}
+
+GhostBoundary::GhostBoundary(const GhostBoundary & other)
+  : RelationshipManager(other),
+    _boundary_name(other._boundary_name),
+    _is_coupling_functor(other._is_coupling_functor),
+    _ghost_point_neighbors(other._ghost_point_neighbors)
+{
+}
+
+void
+GhostBoundary::internalInitWithMesh(const MeshBase &)
+{
+}
+
+std::string
+GhostBoundary::getInfo() const
+{
+  std::ostringstream oss;
+  oss << "GhostBoundary";
+  return oss.str();
+}
+
+void
+GhostBoundary::operator()(const MeshBase::const_element_iterator & /*range_begin*/,
+                          const MeshBase::const_element_iterator & /*range_end*/,
+                          const processor_id_type p,
+                          map_type & coupled_elements)
+{
+  // We ask the user to pass boundary names instead of ids to our constraint object.  However, We
+  // are unable to get the boundary ids from boundary names until we've attached the MeshBase object
+  // to the MooseMesh
+  const bool generating_mesh = !_moose_mesh->getMeshPtr();
+  const auto boundary_ids = generating_mesh ? std::vector<BoundaryID>{Moose::INVALID_BOUNDARY_ID}
+                                            : _moose_mesh->getBoundaryIDs(_boundary_name);
+
+  for (const Elem * const elem : _mesh->active_element_ptr_range())
+  {
+    if (generating_mesh)
+    { // We are still generating the mesh, so it's possible we don't even have the right boundary
+      // ids created yet! So we actually ghost all boundary elements and all lower dimensional
+      // elements who have parents on a boundary
+      if (elem->on_boundary())
+        coupled_elements.insert(std::make_pair(elem, _null_mat));
+      else if (const Elem * const ip = elem->interior_parent())
+      {
+        if (ip->on_boundary())
+          coupled_elements.insert(std::make_pair(elem, _null_mat));
+      }
+    }
+    else
+    {
+      // We've finished generating our mesh so we can be selective and only ghost elements lying
+      // in our lower-dimensional subdomains and their interior parents
+
+      // mooseAssert(boundary_ids != Moose::INVALID_BOUNDARY_ID, "Boundary id should exist by now");
+
+      // Higher-dimensional boundary elements
+      const BoundaryInfo & binfo = _mesh->get_boundary_info();
+
+      for (auto side : elem->side_index_range())
+        for (auto boundary_id : boundary_ids)
+          if ((elem->processor_id() != p) && (binfo.has_boundary_id(elem, side, boundary_id)))
+            coupled_elements.insert(std::make_pair(elem, _null_mat));
```

Comment:
At the point that you've identified the element as needing to be ghosted, you can break out of the boundary id and side loops

Reply:
Breaking out of two loops the easiest way requires a `goto` statement, which is arguably as much bad as it is good. I'm going to suggest that we simply create "a good first" issue to revisit this since you've pointed it out in this review, but I wouldn't necessarily hold up the merge of this PR for this single optimization.

Reply:
People talk about `goto` being bad, but I think it's a dogma. `goto` can be used very effectively and clearly and would be a very reasonable two line addition here to save potentially O(n sides) * O(n boundary IDs) unnecessary looping

Reply:
the amount of time required to implement this change should be under two minutes